vllm/compilation/qk_norm_rope_fusion.py

```diff
@@ -0,0 +1,278 @@
+        positions = empty_i64(T)
+        q_weight = empty_bf16(1, self.head_dim)
+        k_weight = empty_bf16(1, self.head_dim)
+        cos_sin_cache = empty_bf16(4096, self.head_dim)
+        inputs = [
+            qkv,
+            positions,
+            q_weight,
+            k_weight,
+            cos_sin_cache,
+        ]
+
+        # # Register variants across rope ops and with/without contiguous()
+        # # Ensure view ops are canonicalized to reshape in the traced pattern
+        # def wrap_trace_fn(process_fx, trace_fn):
+        #     def wrapped(*args, **kwargs):
+        #         return process_fx(trace_fn(*args, **kwargs))
+
+        #     return wrapped
+
+        # def fx_view_to_reshape(gm: torch.fx.GraphModule):
+        #     from torch._inductor.fx_passes.post_grad import view_to_reshape
+
+        #     view_to_reshape(gm)
+        #     return gm
+
+        pm.register_replacement(
+            pattern,
+            replacement,
+            inputs,
+            pm.fwd_only,
+            pm_pass,
+        )
+
+
+class QKNormRoPEFusionPass(VllmPatternMatcherPass):
+    """Fuse Q/K RMSNorm + RoPE into fused_qk_norm_rope when the custom op exists."""
+
+    @enable_fake_mode
+    def __init__(self, config: VllmConfig):
+        super().__init__(config)
+        self.patterns: PatternMatcherPass = PatternMatcherPass(
+            pass_name="qk_norm_rope_fusion_pass"
+        )
+
+        if not current_platform.is_cuda_alike():
+            logger.debug("QK Norm+RoPE fusion not enabled: unsupported platform")
+            return
+
+        # Register a pattern per attention layer, as sizes differ by shard
+        attn_layers = get_layers_from_vllm_config(config, Attention)
+        if len(attn_layers) == 0:
+            logger.warning(
+                "QK Norm+RoPE fusion enabled, but no Attention layers were discovered."
+            )
+            return
+        layer_name, layer = next(iter(attn_layers.items()))
+
+        for epsilon in [1e-5, 1e-6]:
+            for neox in [True, False]:
+                for rope_op in ROPE_OPS:
+                    try:
+                        QkNormRopePattern(
+                            layer=layer,
+                            eps=epsilon,
+                            rope_op=rope_op,
+                            is_neox=neox,
+                        ).register(self.patterns)
+                    except Exception as e:
+                        logger.debug(
+                            "Skipping QkNormRopePattern register with eps=%s "
+                            "is_neox=%s: %s",
+                            epsilon,
+                            neox,
+                            e,
+                        )
+
+        # Dump patterns for debugging if enabled
+        self.dump_patterns(config, self.patterns)
+
+    @VllmInductorPass.time_and_log
+    def __call__(self, graph: fx.Graph) -> None:
+        if not current_platform.is_cuda_alike():
+            return
```

Comment:
Silent disablement is not good. Should not be enabled at all on non-cuda-alike platforms